File: 2015/ts/day-01/day01.ts
import { readFile } from "fs/promises";

async function calcStairs() {
  let floor = 0;
  let hasReachedBasement = false;
  let position: number;
  const file = await readFile("2015/ts/day-01/input.txt", {
    encoding: "utf-8",
  });
  const data = Buffer.from(file);

  data.forEach((val, index) => {
    if (String.fromCharCode(val) === "(") {
      floor++;
    } else if (String.fromCharCode(val) === ")") {
      floor--;
    }

    if (floor < 0 && !hasReachedBasement) {
      position = index + 1;
      hasReachedBasement = true;
    }
  });
  console.log(floor);
  console.log(position);
}

await calcStairs();

File: 2015/dart/day-01/day-01.dart
import 'dart:io';

void main(List<String> args) async {
  var file = File("2015/dart/day-01/input.txt");
  var floor = 0;
  late int position;
  var hasReachedBasement = false;
  var res = await file.readAsBytes();

  res.asMap().forEach((index, element) {
    if (String.fromCharCode(element).contains("("))
      floor++;
    else if (String.fromCharCode(element).contains(")")) floor--;

    if (floor < 0 && !hasReachedBasement) {
      position = index;
      hasReachedBasement = true;
    }
  });

  print(floor);
  print(position);
}
